include/net/cloud.h

```diff
@@ -0,0 +1,246 @@
+};
+
+/**@brief Structure for cloud backend configuration. */
+struct cloud_backend_config {
+	char *name;
+	cloud_evt_handler_t handler;
+	int socket;
+};
+
+/**@brief Structure for cloud backend. */
+struct cloud_backend {
+	struct cloud_api *api;
+	struct cloud_backend_config *config;
+};
+
+/**@brief API to initialize a cloud backend. Performs all necessary
+ *	  configuration of the backend to be able to connect its online
+ *	  counterpart.
+ *
+ * param[in] cloud Pointer to cloud backend structure.
+ * param[in] ctx Pointer to backend specific data.
+ */
+static inline int cloud_init(
+	struct cloud_backend *const backend,
+	cloud_evt_handler_t handler)
+{
+	if (backend == NULL) {
+		return -EINVAL;
+	}
+
+	if (handler == NULL) {
+		return -EINVAL;
+	}
+
+	return backend->api->init(backend, handler);
+}
+
+/**@brief API to uninitialize a cloud backend. Gracefully disconnects
+ *        remote endpoint and releases memory.
+ *
+ * param[in] cloud Pointer to cloud backend structure.
+ */
+static inline int cloud_uninit(const struct cloud_backend *const backend)
+{
+	__ASSERT_NO_MSG(backend != NULL);
+
+	return backend->api->uninit(backend);
+}
+/**@brief Request connection to cloud backend.
+ *
+ * @details The backend is required to expose the socket in use when this
+ *	    function returns. The socket should then be available through
+ *	    backend->config->socket and the application may start listening
+ *	    for events on it.
+ *
+ * param[in] cloud Pointer to a cloud backend structure.
+ *
+ * @return 0 or a negative error code indicating reason of failure.
+ */
+static inline int cloud_connect(const struct cloud_backend *const backend)
+{
+	__ASSERT_NO_MSG(backend != NULL);
+
+	return backend->api->connect(backend);
+}
+
+/**@brief API to disconnect from cloud backend.
+ *
+ * param[in] cloud Pointer to a cloud backend structure.
+ *
+ * @return 0 or a negative error code indicating reason of failure.
+ */
+static inline int cloud_disconnect(const struct cloud_backend *const backend)
+{
+	__ASSERT_NO_MSG(backend != NULL);
+
+	return backend->api->disconnect(backend);
+}
+
+/**@brief API to send data to cloud.
+ *
+ * param[in] cloud Pointer to a cloud backend structure.
+ * param[in] msg Pointer to cloud message structure.
+ *
+ * @return 0 or a negative error code indicating reason of failure.
+ */
+static inline int cloud_send(const struct cloud_backend *const backend,
+		    struct cloud_msg *msg)
+{
+	__ASSERT_NO_MSG(backend != NULL);
+	__ASSERT_NO_MSG(msg != NULL);
+
+	return backend->api->send(backend, msg);
+}
+
+/**
+ * @brief Optional API to ping the cloud's remote endpoint periodically.
+ *
+ * @param[in] backend Pointer to cloud backend.
+ *
+ * @return 0 or a negative error code indicating reason of failure.
+ */
+static inline int cloud_ping(const struct cloud_backend *const backend)
+{
+	__ASSERT_NO_MSG(backend != NULL);
+
+	/* Ping will only be sent if the backend requires it. */
+	if (backend->api->ping != NULL) {
```

Comment:
How about returning `-ENOTSUP` in case `ping` is not supported?

Reply:
Yep, will fix